File: packages/provide-uterm-server/src/provide/uterm/server/audit_chain.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Tamper-evident WORM audit chain primitive + verifier.

Each appended record carries a monotonic sequence number, a wall-clock and a
monotonic timestamp, and a ``prev_hash`` linking it to the previous record's
``record_hash`` (sha256 over the record's canonical payload). Because every
record's hash covers its predecessor's hash, *any* insertion, deletion,
reordering, or content modification breaks the chain and is detectable by the
verifier.

This is tamper-*evident*, not tamper-*proof*: a writer with the file can still
rewrite the whole log. True immutability requires an append-only sink + an
externally anchored head; that wiring (and cross-instance anchoring) is the
subject of a later sub-task. This module provides only the chain primitive,
the 0600 append-only file sink, and the verification routines.
"""

from __future__ import annotations

import hashlib
import json
import os
import threading
import time
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Callable, Iterable

# The prev_hash of the very first record in a chain (no predecessor). 64 hex
# zeros so it is the same width as a real sha256 digest.
GENESIS_HASH = "0" * 64

# Keys required on a record dict for verification. Mirrors AuditRecord fields.
_RECORD_KEYS = (
    "seq",
    "ts",
    "mono_ns",
    "action",
    "principal",
    "session_id",
    "source_ip",
    "detail",
    "prev_hash",
    "record_hash",
)


def _canonical_payload(
    *,
    seq: int,
    ts: float,
    mono_ns: int,
    action: str,
    principal: str,
    session_id: str,
    source_ip: str,
    detail: dict[str, Any],
    prev_hash: str,
) -> bytes:
    """Deterministic serialization of every field EXCEPT the record's own hash.

    This is THE canonical form: ``compute_record_hash`` digests it, and the
    verifier recomputes it identically from a record's fields. ``prev_hash`` is
    included, so each record's hash chains onto its predecessor.

    ``default=str`` guarantees an exotic value buried in ``detail`` (a set, a
    custom object, ...) can never raise during an audit append — auditing must
    never crash the action it records.
    """
    return json.dumps(
        {
            "seq": seq,
            "ts": ts,
            "mono_ns": mono_ns,
            "action": action,
            "principal": principal,
            "session_id": session_id,
            "source_ip": source_ip,
            "detail": detail,
            "prev_hash": prev_hash,
        },
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        default=str,
    ).encode("utf-8")


def compute_record_hash(payload: bytes) -> str:
    """Return the sha256 hex digest of a canonical payload.

    sha256 is used as an integrity/linking hash, not for password storage.
    """
    return hashlib.sha256(payload).hexdigest()


@dataclass(frozen=True, slots=True)
class AuditRecord:
    """One immutable record in the audit chain."""

    seq: int
    ts: float
    mono_ns: int
    action: str
    principal: str
    session_id: str
    source_ip: str
    detail: dict[str, Any]
    prev_hash: str
    record_hash: str

    def to_dict(self) -> dict[str, Any]:
        """Return the plain dict written as one JSON line (includes record_hash)."""
        return {
            "seq": self.seq,
            "ts": self.ts,
            "mono_ns": self.mono_ns,
            "action": self.action,
            "principal": self.principal,
            "session_id": self.session_id,
            "source_ip": self.source_ip,
            "detail": self.detail,
            "prev_hash": self.prev_hash,
            "record_hash": self.record_hash,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> AuditRecord:
        """Rebuild an AuditRecord from a parsed JSON line."""
        return cls(
            seq=d["seq"],
            ts=d["ts"],
            mono_ns=d["mono_ns"],
            action=d["action"],
            principal=d["principal"],
            session_id=d["session_id"],
            source_ip=d["source_ip"],
            detail=d["detail"],
            prev_hash=d["prev_hash"],
            record_hash=d["record_hash"],
        )


class AuditChain:
    """Append-only, hash-chained audit log backed by a 0600 file."""

    def __init__(
        self,
        path: str | os.PathLike[str],
        *,
        seq: int = 0,
        last_hash: str = GENESIS_HASH,
        clock: Callable[[], float] = time.time,
        mono: Callable[[], int] = time.monotonic_ns,
        on_head: Callable[[int, str], None] | None = None,
    ) -> None:
        """Create a chain writer.

        Parameters
        ----------
        path:
            Destination JSONL file (created 0600 if absent, tightened to 0600
            if it pre-existed with looser perms).
        seq / last_hash:
            Resume state — the head sequence number and hash to continue from.
        clock / mono:
            Injectable wall-clock and monotonic clocks (for deterministic tests).
        on_head:
            Optional callback invoked ``(seq, record_hash)`` after each append;
            a later sub-task passes the control-plane head-persist writer here.
        """
        self._path = os.fspath(path)
        self._seq = seq
        self._last_hash = last_hash
        self._clock = clock
        self._mono = mono
        self._on_head = on_head
        self._lock = threading.Lock()

    @property
    def seq(self) -> int:
        """Current head sequence number."""
        return self._seq

    @property
    def last_hash(self) -> str:
        """Current head record hash."""
        return self._last_hash

    def _write_line(self, line: str) -> None:
        """Append one line to the 0600 file, durably.

        Open-per-append keeps the implementation simple and safe: O_APPEND makes
        each write land atomically at end-of-file even with concurrent writers,
        and the in-process lock keeps seq/last_hash consistent. fsync ensures the
        record reaches disk before the audited action is acknowledged.
        """
        fd = os.open(self._path, os.O_WRONLY | os.O_APPEND | os.O_CREAT, 0o600)
        try:
            # Enforce 0600 even if the file pre-existed with looser perms — an
            # audit log must never be world-readable. fchmod targets the open fd
            # (no TOCTOU on the path).
            os.fchmod(fd, 0o600)
            os.write(fd, line.encode("utf-8"))
            os.fsync(fd)
        finally:
            os.close(fd)

    def append(
        self,
        action: str,
        *,
        principal: str = "",
        session_id: str = "",
        source_ip: str = "",
        detail: dict[str, Any] | None = None,
    ) -> AuditRecord:
        """Append a new audit record, returning it.

        Field names match :func:`provide.uterm.server.audit.audit_event` so a
        later sub-task can wire the two 1:1.
        """
        with self._lock:
            seq = self._seq + 1
            ts = self._clock()
            mono_ns = self._mono()
            detail_d = detail or {}
            prev_hash = self._last_hash
            payload = _canonical_payload(
                seq=seq,
                ts=ts,
                mono_ns=mono_ns,
                action=action,
                principal=principal,
                session_id=session_id,
                source_ip=source_ip,
                detail=detail_d,
                prev_hash=prev_hash,
            )
            record_hash = compute_record_hash(payload)
            record = AuditRecord(
                seq=seq,
                ts=ts,
                mono_ns=mono_ns,
                action=action,
                principal=principal,
                session_id=session_id,
                source_ip=source_ip,
                detail=detail_d,
                prev_hash=prev_hash,
                record_hash=record_hash,
            )
            line = json.dumps(record.to_dict(), separators=(",", ":"), ensure_ascii=False, default=str) + "\n"
            self._write_line(line)
            self._seq = seq
            self._last_hash = record_hash
            if self._on_head is not None:
                self._on_head(seq, record_hash)
            return record

    def anchor(self) -> AuditRecord:
        """Append a checkpoint record snapshotting the head before this record.

        A periodic anchor lets an external notary countersign the head;
        cross-instance/external anchoring is deferred to the HA design.
        """
        anchored_seq = self._seq
        anchored_hash = self._last_hash
        return self.append(
            "audit.anchor",
            detail={"anchored_seq": anchored_seq, "anchored_hash": anchored_hash},
        )

    def close(self) -> None:
        """No-op close (open-per-append keeps no long-lived fd); present for API symmetry."""


@dataclass(frozen=True, slots=True)
class VerifyResult:
    """Outcome of verifying an audit chain."""

    ok: bool
    count: int
    head_seq: int | None
    head_hash: str | None
    first_bad_seq: int | None
    reason: str | None


def verify_records(
    records: Iterable[dict[str, Any]],
    *,
    genesis: str = GENESIS_HASH,
    expected_head: tuple[int, str] | None = None,
) -> VerifyResult:
    """Verify a sequence of record dicts forms an unbroken hash chain.

    The first record establishes the starting sequence number; thereafter the
    sequence must strictly increment by 1. Each record's ``prev_hash`` must
    equal the running hash, and its ``record_hash`` must equal the recomputed
    canonical hash of its own fields. On the FIRST failure a non-ok result is
    returned (never raised). If ``expected_head`` is supplied, the final
    ``(seq, hash)`` must equal it (catches end-truncation / rollback).
    """
    prev = genesis
    expected_seq: int | None = None
    count = 0
    last_seq: int | None = None
    last_hash: str | None = None

    for record in records:
        count += 1
        try:
            for key in _RECORD_KEYS:
                if key not in record:
                    return VerifyResult(False, count, None, None, _seq_of(record), "malformed record")
            seq = record["seq"]
            if not isinstance(seq, int) or isinstance(seq, bool):
                return VerifyResult(False, count, None, None, _seq_of(record), "malformed record")
        except TypeError:
            return VerifyResult(False, count, None, None, None, "malformed record")

        if expected_seq is None:
            expected_seq = seq
        if seq != expected_seq:
            return VerifyResult(False, count, None, None, seq, "non-contiguous sequence")
        if record["prev_hash"] != prev:
            return VerifyResult(False, count, None, None, seq, "broken hash link")

        payload = _canonical_payload(
            seq=seq,
            ts=record["ts"],
            mono_ns=record["mono_ns"],
            action=record["action"],
            principal=record["principal"],
            session_id=record["session_id"],
            source_ip=record["source_ip"],
            detail=record["detail"],
            prev_hash=record["prev_hash"],
        )
        if compute_record_hash(payload) != record["record_hash"]:
            return VerifyResult(False, count, None, None, seq, "record hash mismatch — content altered")

        prev = record["record_hash"]
        last_seq = seq
        last_hash = record["record_hash"]
        expected_seq += 1

    if expected_head is not None and (last_seq, last_hash) != expected_head:
        return VerifyResult(False, count, last_seq, last_hash, last_seq, "head mismatch — log truncated or rolled back")

    return VerifyResult(True, count, last_seq, last_hash, None, None)


def _seq_of(record: Any) -> int | None:
    """Best-effort extraction of a record's seq for error reporting."""
    try:
        value = record.get("seq")
    except AttributeError:
        return None
    return value if isinstance(value, int) and not isinstance(value, bool) else None


def verify_audit_log(
    path: str | os.PathLike[str],
    *,
    expected_head: tuple[int, str] | None = None,
) -> VerifyResult:
    """Verify an on-disk JSONL audit log.

    Reads the file line by line, skipping blank lines; a JSON-decode failure on
    a non-blank line is reported as ``unparseable line N`` (1-based). A missing
    file is reported rather than raised.
    """
    try:
        with Path(path).open(encoding="utf-8") as fh:
            raw_lines = fh.readlines()
    except FileNotFoundError:
        return VerifyResult(False, 0, None, None, None, "audit log not found")

    records: list[dict[str, Any]] = []
    for index, raw in enumerate(raw_lines, start=1):
        line = raw.strip()
        if not line:
            continue
        try:
            records.append(json.loads(line))
        except json.JSONDecodeError:
            return VerifyResult(False, len(records), None, None, index, f"unparseable line {index}")

    return verify_records(records, expected_head=expected_head)

File: packages/provide-uterm-go/cli/auditchain.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package cli

import (
	"bufio"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"strings"
)

// genesisHash is the prev_hash of the very first record in a chain (64 hex
// zeros — same width as a real sha256 digest). Mirrors audit_chain.GENESIS_HASH.
const genesisHash = "0000000000000000000000000000000000000000000000000000000000000000"

// recordKeys are the fields required on every audit record. Mirrors _RECORD_KEYS.
var recordKeys = []string{
	"seq", "ts", "mono_ns", "action", "principal",
	"session_id", "source_ip", "detail", "prev_hash", "record_hash",
}

// payloadKeys are the fields covered by the record hash (every field EXCEPT the
// record's own hash). Mirrors _canonical_payload.
var payloadKeys = []string{
	"seq", "ts", "mono_ns", "action", "principal",
	"session_id", "source_ip", "detail", "prev_hash",
}

// verifyResult is the outcome of verifying an audit chain. Mirrors VerifyResult.
type verifyResult struct {
	OK          bool
	Count       int
	HeadSeq     *int64
	HeadHash    *string
	FirstBadSeq *int64
	Reason      string
}

// expectedHead is an optional (seq, hash) head assertion.
type expectedHead struct {
	seq  int64
	hash string
}

// computeRecordHash returns the sha256 hex digest of a canonical payload.
func computeRecordHash(payload []byte) string {
	sum := sha256.Sum256(payload)
	return hex.EncodeToString(sum[:])
}

// canonicalPayload deterministically serializes every payload field of a parsed
// record (record_hash excluded) exactly as CPython's
//
//	json.dumps(subset, sort_keys=True, separators=(",",":"), ensure_ascii=False, default=str)
//
// would. This is THE parity surface: the record hash is a sha256 over these
// bytes, so a single byte of divergence makes a valid log look tampered.
//
// Parity resolution (see auditchain_test.go differential test): the Python
// chain does NOT hash the raw JSONL line — it re-serializes a sorted subset. But
// json.dumps of a value parsed back from the line is byte-identical to the
// line's own rendering (shortest-round-trip float repr is stable), so numbers
// are reproduced by parsing them as json.Number and re-rendering through Python's
// repr() algorithm (pyAuditFloatRepr). default=str never fires here because a
// parsed JSONL value is already JSON-native.
func canonicalPayload(record map[string]any) ([]byte, error) {
	subset := make(map[string]any, len(payloadKeys))
	for _, k := range payloadKeys {
		subset[k] = record[k]
	}
	var b strings.Builder
	if err := auditEncode(&b, subset); err != nil {
		return nil, err
	}
	return []byte(b.String()), nil
}

// verifyAuditLog verifies an on-disk JSONL audit log. Mirrors verify_audit_log.
func verifyAuditLog(path string, head *expectedHead) verifyResult {
	f, err := os.Open(path) //nolint:gosec // path is an operator-supplied audit log
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return verifyResult{Reason: "audit log not found"}
		}
		return verifyResult{Reason: "audit log not found"}
	}
	defer f.Close() //nolint:errcheck // read-only file

	var records []map[string]any
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 64*1024), 16*1024*1024)
	lineNo := 0
	for scanner.Scan() {
		lineNo++
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		rec, perr := parseRecordLine(line)
		if perr != nil {
			bad := int64(lineNo)
			return verifyResult{Count: len(records), FirstBadSeq: &bad, Reason: fmt.Sprintf("unparseable line %d", lineNo)}
		}
		records = append(records, rec)
	}
	if err := scanner.Err(); err != nil {
		return verifyResult{Reason: "audit log not found"}
	}
	return verifyRecords(records, head)
}

// parseRecordLine decodes one JSONL line into a map, preserving number tokens
// as json.Number (Python's json.loads keeps the int/float distinction, which is
// load-bearing for canonical re-serialization).
func parseRecordLine(line string) (map[string]any, error) {
	dec := json.NewDecoder(strings.NewReader(line))
	dec.UseNumber()
	var v any
	if err := dec.Decode(&v); err != nil {
		return nil, err
	}
	if dec.More() {
		return nil, errors.New("trailing data")
	}
	m, ok := v.(map[string]any)
	if !ok {
		return nil, errors.New("record is not an object")
	}
	return m, nil
}

// verifyRecords verifies a sequence of parsed records forms an unbroken hash
// chain. Mirrors verify_records: the first record establishes the starting
// sequence; thereafter seq must strictly increment by 1, each prev_hash must
// equal the running hash, and each record_hash must equal the recomputed
// canonical hash. On the FIRST failure a non-ok result is returned.
func verifyRecords(records []map[string]any, head *expectedHead) verifyResult {
	prev := genesisHash
	var expectedSeq int64
	haveExpected := false
	count := 0
	var lastSeq *int64
	var lastHash *string

	for _, record := range records {
		count++
		if res, bad := checkRecordShape(record, count); bad {
			return res
		}
		seq, _ := recordSeq(record)
		if !haveExpected {
			expectedSeq = seq
			haveExpected = true
		}
		if seq != expectedSeq {
			s := seq
			return verifyResult{Count: count, FirstBadSeq: &s, Reason: "non-contiguous sequence"}
		}
		if asStr(record["prev_hash"]) != prev {
			s := seq
			return verifyResult{Count: count, FirstBadSeq: &s, Reason: "broken hash link"}
		}
		payload, err := canonicalPayload(record)
		if err != nil {
			s := seq
			return verifyResult{Count: count, FirstBadSeq: &s, Reason: "malformed record"}
		}
		if computeRecordHash(payload) != asStr(record["record_hash"]) {
			s := seq
			return verifyResult{Count: count, FirstBadSeq: &s, Reason: "record hash mismatch — content altered"}
		}
		rh := asStr(record["record_hash"])
		prev = rh
		sc := seq
		lastSeq, lastHash = &sc, &rh
		expectedSeq++
	}

	if head != nil {
		match := lastSeq != nil && lastHash != nil && *lastSeq == head.seq && *lastHash == head.hash
		if !match {
			return verifyResult{
				Count: count, HeadSeq: lastSeq, HeadHash: lastHash, FirstBadSeq: lastSeq,
				Reason: "head mismatch — log truncated or rolled back",
			}
		}
	}
	return verifyResult{OK: true, Count: count, HeadSeq: lastSeq, HeadHash: lastHash}
}

// checkRecordShape validates a record has all required keys and an integer seq.
func checkRecordShape(record map[string]any, count int) (verifyResult, bool) {
	for _, key := range recordKeys {
		if _, ok := record[key]; !ok {
			return verifyResult{Count: count, FirstBadSeq: seqOf(record), Reason: "malformed record"}, true
		}
	}
	if _, ok := recordSeq(record); !ok {
		return verifyResult{Count: count, FirstBadSeq: seqOf(record), Reason: "malformed record"}, true
	}
	return verifyResult{}, false
}

// recordSeq extracts an integer seq. A float/exponential/boolean seq is not an
// integer (mirrors Python's isinstance(seq, int) and not isinstance(seq, bool)).
func recordSeq(record map[string]any) (int64, bool) {
	n, ok := record["seq"].(json.Number)
	if !ok {
		return 0, false
	}
	s := string(n)
	if strings.ContainsAny(s, ".eE") {
		return 0, false
	}
	v, err := n.Int64()
	if err != nil {
		return 0, false
	}
	return v, true
}

// seqOf best-effort extracts a record's seq for error reporting.
func seqOf(record map[string]any) *int64 {
	if v, ok := recordSeq(record); ok {
		return &v
	}
	return nil
}

func asStr(v any) string {
	s, _ := v.(string)
	return s
}
